### PR TITLE
add plugin support oc prefix

### PR DESCRIPTION
**Pull request by juanvallejo**
Will open upstream PR once we reach some general consensus on this approach
~~Work in progress, will need to also patch the `plugin list` command.~~

cc @soltysh

### CODE REVIEW

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/cmd.go [hunk 1]
```diff
@@ -320,7 +324,7 @@ func (h *defaultPluginHandler) Execute(executablePath string, cmdArgs, environme
 	return syscall.Exec(executablePath, cmdArgs, environment)
 }
 
-func handleEndpointExtensions(pluginHandler PluginHandler, cmdArgs []string) error {
+func HandlePluginCommand(pluginHandler PluginHandler, validPrefixes, cmdArgs []string) error {
```
Comment (soltysh) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/cmd.go]: Move the prefixes to `NewDefaultPluginHandler` constructor.

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/cmd.go [hunk 2]
```diff
@@ -304,6 +304,10 @@ type PluginHandler interface {
 
 type defaultPluginHandler struct{}
 
+func NewDefaultPluginHandler() *defaultPluginHandler {
```
Comment (soltysh) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/cmd.go]: Use this constructor in `NewDefaultKubectlCommand` too.

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/cmd.go [hunk 3]
```diff
@@ -334,8 +338,8 @@ func handleEndpointExtensions(pluginHandler PluginHandler, cmdArgs []string) err
 
 	// attempt to find binary, starting at longest possible name with given cmdArgs
 	for len(remainingArgs) > 0 {
-		path, err := pluginHandler.Lookup(fmt.Sprintf("kubectl-%s", strings.Join(remainingArgs, "-")))
```
Comment (soltysh) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/cmd.go]: With `prefixes` being part of `defaultPluginHandler` struct you won't have to change this bit.

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/cmd.go [hunk 4]
```diff
@@ -358,6 +362,17 @@ func handleEndpointExtensions(pluginHandler PluginHandler, cmdArgs []string) err
 	return nil
 }
 
+func lookupPluginFilenameByPrefix(pluginHandler PluginHandler, validPrefixes []string, potentialFilename string) string {
```
Comment (soltysh) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/cmd.go]: This shouldn't be needed either.

File: pkg/oc/cli/cli.go [hunk 5]
```diff
@@ -330,14 +330,14 @@ func CommandFor(basename string) *cobra.Command {
...
-		cmd = kubecmd.NewKubectlCommand(in, out, errout)
+		cmd = kubecmd.NewDefaultKubectlCommand()
 	case "openshift-deploy":
 		cmd = deployer.NewCommandDeployer(basename)
 	case "openshift-recycle":
 		cmd = recycle.NewCommandRecycle(basename, out)
 	default:
 		shimKubectlForOc()
-		cmd = NewCommandCLI("oc", "oc", in, out, errout)
+		cmd = withPluginsSupport(NewCommandCLI("oc", "oc", in, out, errout), os.Args, out, errout)
```
Comment (soltysh) [inline pkg/oc/cli/cli.go]: Make a wrapper similar to `NewDefaultKubectlCommandWithArgs` and use it here without the current nasty function in function.

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/plugin.go [hunk 6]
```diff
@@ -120,12 +124,12 @@ func (o *PluginListOptions) Run() error {
...
 				continue
 			}
-			if !strings.HasPrefix(f.Name(), "kubectl-") {
+			if !hasValidPrefix(f.Name(), ValidPluginFilenamePrefixes) {
 				continue
 			}
 
 			if isFirstFile {
-				fmt.Fprintf(o.ErrOut, "The following kubectl-compatible plugins are available:\n\n")
+				fmt.Fprintf(o.ErrOut, "The following plugins are available:\n\n")
```
Comment (soltysh) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/plugin.go]: leave the word `compatible` here

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/cmd.go (hunk 2 above)
Comment (soltysh) [reply vendor/k8s.io/kubernetes/pkg/kubectl/cmd/cmd.go]: Other option is to use a global var similar to `ValidPluginFilenamePrefixes` and use it in both needed places.

Comment (soltysh) [reply vendor/k8s.io/kubernetes/pkg/kubectl/cmd/cmd.go]: Yeah, I'm looking into that and just re-using that same constant here will be the simplest approach and will be also easier to justify upstream :wink:

**Comment by juanvallejo** [comment]
@soltysh thanks, comments addressed

File: pkg/oc/cli/cli.go [hunk 7]
```diff
@@ -330,14 +352,14 @@ func CommandFor(basename string) *cobra.Command {
...
 	case "kubectl":
-		cmd = kubecmd.NewKubectlCommand(in, out, errout)
+		cmd = kubecmd.NewDefaultKubectlCommand()
 	case "openshift-deploy":
 		cmd = deployer.NewCommandDeployer(basename)
 	case "openshift-recycle":
 		cmd = recycle.NewCommandRecycle(basename, out)
 	default:
 		shimKubectlForOc()
-		cmd = NewCommandCLI("oc", "oc", in, out, errout)
```
Comment (soltysh) [inline pkg/oc/cli/cli.go]: While you're at it rename this to be `NewDefaultOcCommand` and the next to `NewOcCommand`

**Pull request merged**